site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,685 @@
+	Rules []GRPCRouteRule `json:"rules,omitempty"`
+}
+
+// GRPCRouteRule defines semantics for matching an gRPC request based on
+// conditions (matches), processing it (filters), and forwarding the request to
+// an API object (backendRefs).
+type GRPCRouteRule struct {
+	// Matches define conditions used for matching the rule against incoming
+	// gRPC requests. Each match is independent, i.e. this rule will be matched
+	// if **any** one of the matches is satisfied.
+	//
+	// For example, take the following matches configuration:
+	//
+	// ```
+	// matches:
+	// - method:
+	//     method: foo.bar
+	//   headers:
+	//     values:
+	//       version: 2
+	// - method:
+	//     service: foo.bar.v2
+	// ```
+	//
+	// For a request to match against this rule, a request should satisfy
+	// EITHER of the two conditions:
+	//
+	// - service of foo.bar AND contains the header `version: 2`
+	// - service of foo.bar.v2
+	//
+	// See the documentation for GRPCRouteMatch on how to specify multiple
+	// match conditions that should be ANDed together.
+	//
+	// If no matches are specified, the implementation must match every gRPC request.
+	//
+	// Proxy or Load Balancer routing configuration generated from GRPCRoutes
+	// MUST prioritize rules based on the following criteria, continuing on
+	// ties. Merging must not be done between GRPCRoutes and HTTPRoutes.
+	// Precedence must be given to the rule with the largest number of:
+	//
+	// * Characters in a matching non-wildcard hostname.
+	// * Characters in a matching hostname.
+        // * Characters in a matching service.
+        // * Characters in a matching method.
+	// * Header matches.
+	//
+	// If ties still exist across multiple Routes, matching precedence MUST be
+	// determined in order of the following criteria, continuing on ties:
+	//
+	// * The oldest Route based on creation timestamp.
+	// * The Route appearing first in alphabetical order by
+	//   "{namespace}/{name}".
+	//
+	// If ties still exist within the Route that has been given precedence,
+	// matching precedence MUST be granted to the first matching rule meeting
+	// the above criteria.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{method: {type: "Exact"}}}
+	Matches []GRPCRouteMatch `json:"matches,omitempty"`
+
+	// Filters define the filters that are applied to requests that match
+	// this rule.
+	//
+	// The effects of ordering of multiple behaviors are currently unspecified.
+	// This can change in the future based on feedback during the alpha stage.
+	//
+	// Conformance-levels at this level are defined based on the type of filter:
+	//
+	// - ALL core filters MUST be supported by all implementations.
+	// - Implementers are encouraged to support extended filters.
+	// - Implementation-specific custom filters have no API guarantees across
+	//   implementations.
+	//
+	// Specifying a core filter multiple times has unspecified or custom
+	// conformance.
+	// Support: Core
```

Comment:
This has come before so I want to clarify.
Since gRPC filters are part of a bigger feature, gRPCRoute, they shouldn't have a conformance level that is higher than the parent.

Should this be `extended`?

Reply:
I think I need some guidance here.

> Since gRPC filters are part of a bigger feature, gRPCRoute, they shouldn't have a conformance level that is higher than the parent.

I absolutely agree with this. In the places where I've written "Core" in this proposal, I've meant "Core to GRPCRoute."  Likewise, "Custom" here means "Custom to GRPCRoute."

If I write the support levels in absolute terms, then I think that would flatten things out and everything would become "Support: extended." This loses much of the detail that I've written in about support levels between different GRPCRoute implementations.

Thoughts on how to resolve this?

Reply:
> I absolutely agree with this. In the places where I've written "Core" in this proposal, I've meant "Core to GRPCRoute." Likewise, "Custom" here means "Custom to GRPCRoute."

That is how I read this as well but we run into issues such as #655.

> If I write the support levels in absolute terms, then I think that would flatten things out and everything would become "Support: extended." This loses much of the detail that I've written in about support levels between different GRPCRoute implementations.

Extended features must be portable _if_ they are supported. Is that not enough for defining semantics here?

cc @robscott @youngnick

Reply:
I don't think we've explicitly discussed this before, but I've always conceived of the field-level support working like @gnossen described - this is because we clearly _have_ to allow implementations not to support some objects, because Layer 7 ingress controllers may not be able to do Layer 4 resources (like TCPRoute and UDPRoute), and Layer 4 load balancers may not be able to do higher-layer resources (like HTTPRoute or TLSRoute).

So, the "Support: Core" markings on fields have to be on a per-object basis.

But like I said, this is a subtle point that I don't think we've explicitly made before.

Reply:
I agree with the interpretation here. We do cover this in the docs, but it may not be prominent enough (https://gateway-api.sigs.k8s.io/concepts/guidelines/#overlapping-support-levels).